### DISCUSSION TITLE
Multiple base Profiles

### DISCUSSION THREAD

**Discussion by Felix-Braun** [zowe/zowe-cli#2114] [Q&A]
Hello,

I wanted to ask, if there is the possibility of creating multiple Base Profiles, to use.
For example, one with Authentication Token and one with User and Password. And then to specify in the profiles itself which to use?

Thanks in advance. :)

**Comment by adam-wolfe** [comment]
Hi @Felix-Braun, this particular scenario should be possible using [profile nesting](https://docs.zowe.org/stable/user-guide/cli-using-team-configuration-team-leaders/#access-to-one-or-more-lpars-contain-services-that-do-not-share-the-same-credentials) in Zowe CLI/Explorer V2. 

As an example from a config I use for demos and development, I have a demo system that I connect to with user name and password, then I have a dev system that authenticates through the APIML. The "demo" profile includes all properties needed to connect to a given LPAR which are then used for the "demo.zosmf" profile, while the "dev.zosmf" profile relies on the properties stored in the base profile:

```
{
    "$schema": "./zowe.schema.json",
    "profiles": {
        "demo": {
            "properties": {
                "host": "demo.mainframe.net"
            },
            "secure": [
                "user",
                "password"
            ],
            "profiles": {
                "zosmf": {
                    "type": "zosmf",
                    "properties": {
                        "port": 1234
                    }
                }
            }
        },
        "dev": {
            "profiles": {
                "zosmf": {
                    "type": "zosmf",
                    "properties": {
                        "basePath": "/ibmzosmf/api/v1"
                    }
                }
            }
        },
        "base_apiml": {
            "type": "base",
            "properties": {
                "host": "dev.mainframe.net",
                "port": 10000,
                "tokenType": "apimlAuthenticationToken"
            },
            "secure": [
                "tokenValue"
            ]
        }
    },
    "defaults": {
        "zosmf": "dev.zosmf",
        "base": "base_apiml"
    }
}
```

There are some limitations with this if you are connecting to multiple instances of the APIML (because APIML connections must be defined within a base profile, and you can only have one default base profile). However, if APIML is not involved, you could use nesting to connect to any number of systems using different authentication methods.

For V3, we are planning to allow users to specify profile links that could link service profiles to a base profile. This is related to issue: https://github.com/zowe/zowe-cli/issues/1705, however, I think we are considering explicit links between profiles rather than relying only on a nested config structure.

**Comment by Felix-Braun** [comment]
Hello @adam-wolfe, i just made the nested profile:

```
{
    "$schema": "./zowe.schema.json",
    "profiles": {
        "userpwd": {
            "properties": {
                "host": "changeman-rest.net",
                "basePath": "/zmfrest",
                "port": 443,
                "protocol": "https"
            },
            "profiles": {
                "zmf1": {
                    "type": "zmf",
                    "properties": {
                        "port": 443,
                        "protocol": "https",
                        "encoding": "IBM-1141",
                        "rejectUnauthorized": false
                    }
                }
            },
            "secure": [
                "user",
                "password"
            ]
        },
        "apiml": {
            "profiles": {
                "zosmf": {
                    "type": "zosmf",
                    "properties": {
                        "basePath": "ibmzosmf/api/v1",
                        "encoding": "IBM-1141",
                        "responseTimeout": 300,
                        "port": 0,
                        "tokenType": "apimlAuthenticationToken"
                    },
                    "secure": [
                        "tokenValue"
                    ]
                }
            }
        },
        "base": {
            "type": "base",
            "properties": {
                "host": "gateway-zowe.net",
                "rejectUnauthorized": false,
                "tokenType": "apimlAuthenticationToken"
            },
            "secure": [
                "tokenValue",
                "user",
                "password"
            ]
        }
    },
    "defaults": {
        "zosmf": "apiml.zosmf",
        "zmf": "userpwd.zmf1",
        "base": "base"
    },
    "autoStore": true
}
``` 

Everytime I want to use the Profile with the Username and Password, it add the user and Password to the Baseprofile, after that, if I try to logon through apiml, it asks me for Username and Password.

And in the CLI, it doesnt give me an error, it just does nothing. I don't know if you know ZMF extension/ CLI Plugin, otherwise I'll ask them to help me.

Could this be a Problem from ZMF-Extensions? 

Thanks in advance.

**Comment by gejohnston** [comment]
Did you place the properties tokenType, tokenValue, user, and password into your profiles by hand in your text editor?

The reason that I ask is that Zowe CLI software would not place tokenType and tokenValue into your apiml.profiles.zosmf profile. Also, it would not leave both user/password, and tokenType/TokenValue in your base profile.

If my supposition is correct, I suggest that you create a clean slate by doing the following, and then make Zowe CLI create the correct token-related properties for you.

- In your apiml.profiles.zosmf profile:
    - Remove tokenType and tokenValue.
    - Because this profile has a basePath property, it will automatically get tokenValue and TokenType from your base profile.

- In your base profile:
    - Add a port value for your instance of APIML. Otherwise Zowe will not be able to connect to APIML.
    - Remove tokenType.
    - Remove tokenValue, user, and password from the secure array.

If you want to understand any of the upcoming commands, you can get descriptions of them with the following command:

```
zowe --help-web
```

- If you have not securely stored your user and password for userpwd.zmf1 after changing your original configuration, do so now with the following command. That command will prompt you for all secure properties and it will securely store your responses. At this point you should only be prompted for the user and password from userpwd.zmf1. Use the --global-config option (or not) depending on where you have placed your zowe.config.json file.

```
zowe config secure --global-config
```

- Run the following Zowe CLI command. It will ask you for the user and password to authenticate to your APIML server. That command will **NOT** store your user and password. It will obtain a token from APIML, and store the right token properties for you in your base profile (and tokenValue in the secure vault).

```
zowe auth login apiml
```

You should now have a correct configuration. Test your apiml.zosmf profile with a simple CLI command, like the following. If that command works you can then try operations with the zmf plugin. You may need to follow up with your provider of the zmf plugin if you then have difficulties. 

```
zowe zos-files list data-set "SYS1.PARMLIB*"
```
